File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/CdmFolder/EntityReferencePersistence.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.CdmFolder
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Newtonsoft.Json.Linq;
    using System.Collections.Generic;

    class EntityReferencePersistence
    {
        public static CdmEntityReference FromData(CdmCorpusContext ctx, JToken obj)
        {
            if (obj == null)
            {
                return null;
            }

            dynamic entity;
            bool simpleReference = true;

            if (obj is JValue)
            {
                entity = (string)obj;
            }
            else
            {
                entity = GetEntityReference(ctx, obj);
                simpleReference = false;
            }

            CdmEntityReference entityReference = ctx.Corpus.MakeRef<CdmEntityReference>(CdmObjectType.EntityRef, entity, simpleReference);

            if (!(obj is JValue))
            {
                Utils.AddListToCdmCollection(entityReference.AppliedTraits, Utils.CreateTraitReferenceList(ctx, obj["appliedTraits"]));
            }

            return entityReference;
        }

        public static dynamic ToData(CdmEntityReference instance, ResolveOptions resOpt, CopyOptions options)
        {
            if (instance.ExplicitReference != null && instance.ExplicitReference.GetType() == typeof(CdmProjection))
            {
                return ProjectionPersistence.ToData(instance.ExplicitReference as CdmProjection, resOpt, options);
            }
            else
            {
                return CdmObjectRefPersistence.ToData(instance, resOpt, options);
            }
        }

        private static dynamic GetEntityReference(CdmCorpusContext ctx, JToken obj)
        {
            dynamic entity = null;
            if (obj["entityReference"] is JValue)
                entity = obj["entityReference"];
            else if (obj["entityReference"]?["entityShape"] != null)
                entity = ConstantEntityPersistence.FromData(ctx, obj["entityReference"]);
            else if (obj["source"] != null || obj["operations"] != null)
                entity = ProjectionPersistence.FromData(ctx, obj);
            else
                entity = EntityPersistence.FromData(ctx, obj["entityReference"]);
            return entity;
        }
    }
}

File: objectModel/Python/cdm/persistence/cdmfolder/entity_reference_persistence.py
﻿# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

from typing import Union, Optional, cast
from cdm.enums import CdmObjectType
from cdm.objectmodel import CdmCorpusContext, CdmEntityReference, CdmProjection

from . import utils
from .cdm_object_ref_persistence import CdmObjectRefPersistence
from .constant_entity_persistence import ConstantEntityPersistence
from .projections.projection_persistence import ProjectionPersistence
from .types import EntityReference


class EntityReferencePersistence(CdmObjectRefPersistence):
    @staticmethod
    def from_data(ctx: CdmCorpusContext, data: Union[str, EntityReference]) -> Optional[CdmEntityReference]:
        if not data:
            return None

        simple_reference = True
        entity = None

        if isinstance(data, str):
            entity = data
        else:
            entity = EntityReferencePersistence._get_entity_reference(ctx, data)
            simple_reference = False

        entity_reference = ctx.corpus.make_ref(CdmObjectType.ENTITY_REF, entity, simple_reference)

        if not isinstance(data, str) and entity_reference:
            utils.add_list_to_cdm_collection(entity_reference.applied_traits,
                                             utils.create_trait_reference_array(ctx, data.get('appliedTraits')))

        return entity_reference

    @staticmethod
    def to_data(instance: 'CdmEntityReference', res_opt: 'ResolveOptions', options: 'CopyOptions') -> 'EntityReference':
        if not instance:
            return None

        if instance.explicit_reference and isinstance(instance.explicit_reference, CdmProjection):
            return ProjectionPersistence.to_data(cast('CdmProjection', instance.explicit_reference), res_opt, options)
        else:
            return CdmObjectRefPersistence.to_data(instance, res_opt, options)

    @staticmethod
    def _get_entity_reference(ctx: 'CdmCorpusContext', data) -> 'CdmObject':
        from .entity_persistence import EntityPersistence

        entity = None
        if isinstance(data.entityReference, str):
            entity = data.entityReference
        elif data.entityReference and data.entityReference.get('entityShape'):
            entity = ConstantEntityPersistence.from_data(ctx, data.entityReference)
        elif data.get('source') or data.get('operations'):
            entity = ProjectionPersistence.from_data(ctx, data)
        else:
            entity = EntityPersistence.from_data(ctx, data.entityReference)
        return entity
